File: javascript/breadthFirstSearch.js
"use strict";

const friends = {};
friends.me = ["tom", "dick", "harry"];
friends.tom = ["me", "dick"];
friends.dick = ["me", "tom"];
friends.harry = ["me"];

console.log(friends);

const friendHasM = (name) => name[name.length - 1] === "m";

const bfs = (name) => {
  let searchFriends = [];
  searchFriends = searchFriends.concat(friends[name]);
  const searched = [];
  while (searchFriends.length) {
    const friend = searchFriends.shift();
    if (searched.indexOf(friend) === -1) {
      if (friendHasM(friend)) {
        console.log(`${friend}'s name ends in the letter 'm'`);
        return true;
      }
      searchFriends = searchFriends.concat(friends[friend]);
      searched.push(friend);
    }
  }
  return false;
};

bfs("me");

File: python/breadth_first_search.py
from collections import deque

friends = {}

friends["me"] = ["tom", "dick", "harry"]
friends["tom"] = ["me", "dick"]
friends["dick"] = ["me", "tom"]
friends["harry"] = ["me"]

print(friends)


def friend_has_m(name):
    return name[-1] == 'm'


def bfs(name):
    search_friends = deque()
    search_friends += friends[name]
    searched = []
    while search_friends:
        friend = search_friends.popleft()
        if friend not in searched:
            if friend_has_m(friend):
                print friend + "'s name ends in the letter 'm'"
                return True
            else:
                search_friends += friends[friend]
                searched.append(friend)
    return False


print(bfs("me"))

